File: onnx2trt/C++/export_trt.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <NvInfer.h>
#include <NvOnnxParser.h>

#include "utils.h"

const int nHeight = 28;
const int nWidth = 28;
const std::string onnx_path = "../model.onnx";
const std::string trtFile = "./model.plan";
const std::string dataFile = "./data.npz";
static onnx2trt::Logger gLogger(nvinfer1::ILogger::Severity::kERROR);

// for FP16 mode
const bool bFP16Mode =true;
// for INT8 mode
const bool bINT8Mode = false;
const int nCalibration = 1;
const std::string cacheFile = "./int8.cache";
const std::string calibrationDataFile = std::string("./data.npz");

int main()
{
    CUDA_CHECK(cudaSetDevice(0));
    nvinfer1::ICudaEngine *engine = nullptr;

    nvinfer1::IBuilder *builder = nvinfer1::createInferBuilder(gLogger);
    nvinfer1::INetworkDefinition *network = builder->createNetworkV2(
        1U << int(nvinfer1::NetworkDefinitionCreationFlag::kEXPLICIT_BATCH)
    );
    nvinfer1::IOptimizationProfile *profile = builder->createOptimizationProfile();
    nvinfer1::IBuilderConfig *config = builder->createBuilderConfig();
    nvinfer1::IInt8Calibrator *pCalibrator = nullptr;

    if (bFP16Mode)
    {
        config->setFlag(nvinfer1::BuilderFlag::kFP16);
    }
    // if (bINT8Mode)
    // {
    //     config->setFlag(nvinfer1::BuilderFlag::kINT8);
    //     nvinfer1::Dims inputShape {4, {1, 1, nHeight, nWidth}};
    //     pCalibrator = new MyCalibrator(calibrationDataFile, nCalibration, inputShape, cacheFile);
    //     if (pCalibrator == nullptr)
    //     {
    //         std::cout << std::string("Failed getting Calibrator for Int8!") << std::endl;
    //         return 1;
    //     }
    //     config->setInt8Calibrator(pCalibrator);
    // }

    nvonnxparser::IParser *parser = nvonnxparser::createParser(*network, gLogger);
    if (!parser->parseFromFile(onnx_path.c_str(), int(gLogger.getSeverity())))
    {
        std::cout << std::string("Failed parsing .onnx file!") << std::endl;
        for (int i = 0; i < parser->getNbErrors(); ++i)
        {
            auto *error = parser->getError(i);
            std::cout << std::to_string(int(error->code())) << std::string(":") << std::string(error->desc()) << std::endl;
        }
        return 1;
    }
    std::cout << std::string("Succeeded parsing .onnx file!") << std::endl;

    nvinfer1::ITensor *inputTensor = network->getInput(0);
    profile->setDimensions(inputTensor->getName(), nvinfer1::OptProfileSelector::kMIN, nvinfer1::Dims {4, {1, 1, nHeight, nWidth}});
    profile->setDimensions(inputTensor->getName(), nvinfer1::OptProfileSelector::kOPT, nvinfer1::Dims {4, {4, 1, nHeight, nWidth}});
    profile->setDimensions(inputTensor->getName(), nvinfer1::OptProfileSelector::kMAX, nvinfer1::Dims {4, {8, 1, nHeight, nWidth}});
    config->addOptimizationProfile(profile);

    network->unmarkOutput(*network->getOutput(0)); // remove output tensor "y"

    nvinfer1::IHostMemory *engineString = nullptr;
#if TENSORRT_VERSION == 8
    engineString = builder->buildSerializedNetwork(*network, *config);
#else
    engine = builder->buildCudaEngine(*network);
    engineString = engine->serialize();
#endif

    if (engineString == nullptr || engineString->size() == 0)
    {
        std::cout << "Failed building serialized engine!" << std::endl;
        return 1;
    }
    std::cout << "Succeeded building serialized engine!" << std::endl;

    nvinfer1::IRuntime *runtime = nvinfer1::createInferRuntime(gLogger);
    engine = runtime->deserializeCudaEngine(engineString->data(), engineString->size());
    if (engine == nullptr)
    {
        std::cout << "Failed building engine!" << std::endl;
        return 1;
    }
    std::cout << "Succeeded building engine!" << std::endl;

    if (bINT8Mode && pCalibrator != nullptr)
    {
        delete pCalibrator;
    }

    // save engine to file
    std::ofstream engineFile(trtFile, std::ios::binary);
    if (!engineFile)
    {
        std::cout << "Failed opening file to write" << std::endl;
        return 1;
    }
    engineFile.write(static_cast<char *>(engineString->data()), engineString->size());
    if (engineFile.fail())
    {
        std::cout << "Failed saving .plan file!" << std::endl;
        return 1;
    }
    std::cout << "Succeeded saving .plan file!" << std::endl;

    return 0;
}

File: onnx2trt/export_trt.py
import os
import argparse

import tensorrt as trt
from cuda import cudart

from core.calibrator import MyCalibrator


def make_parser():
    parser = argparse.ArgumentParser("Export TensorRT engine from Onnx")
    parser.add_argument("--onnx", required=True, help="onnx model path")
    parser.add_argument("-output", required=True, help="output TensorRT engine path")
    parser.add_argument(
        "--shape", required=True, help="model input shape with format `C,H,W`"
    )
    parser.add_argument(
        "--fp16", action="store_true", help="Whether to enable fp16 mode"
    )
    parser.add_argument(
        "--int8", action="store_true", help="Whether to enable int8 mode"
    )
    parser.add_argument(
        "--calibration_data", default=None, help="Calibration data path for int8"
    )
    parser.add_argument(
        "--calibration_cache", default=None, help="Calibration cache path for int8"
    )
    return parser


def export(
    onnx_path: str,
    output_path: str,
    input_shape: tuple,
    precision_mode="fp16",
    calibrator=None,
):
    """
    Args:
        onnx_path: onnx path
        input_shape: preferred input size of model with format (C, H, W)
    """
    logger = trt.Logger(trt.Logger.VERBOSE)
    builder = trt.Builder(logger)
    network = builder.create_network(
        1 << int(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)
    )
    profile = builder.create_optimization_profile()
    config = builder.create_builder_config()

    if precision_mode == "fp16":
        config.set_flag(trt.BuilderFlag.FP16)
    elif precision_mode == "int8":
        config.set_flag(trt.BuilderFlag.INT8)
        config.int8_calibrator = calibrator
    else:
        raise NotImplementedError(f"precision mode {precision_mode} not implemented.")

    if not os.path.exists(onnx_path):
        print(f"ONNX file {onnx_path} not exists.")
        return False

    print("loading ONNX file...")
    parser = trt.OnnxParser(network, logger)
    # parse_valid = parser.parse_from_file(onnx_path)
    with open(onnx_path, "rb") as model:
        if not parser.parse(model.read()):
            print("Failed parsing .onnx file!")
            for error in range(parser.num_errors):
                print(parser.get_error(error))
            return False
        print("Succeeded parsing .onnx file!")

    input_tensor = network.get_input(0)
    profile.set_shape(
        input_tensor.name,
        [1, *input_shape],
        [4, *input_shape],
        [8, *input_shape],
    )
    config.add_optimization_profile(profile)

    engine_string = builder.build_serialized_network(network, config)
    if engine_string == None:
        print("Failed building engine!")
        return False

    print("Succeeded building engine!")
    with open(output_path, "wb") as f:
        f.write(engine_string)

    return True


if __name__ == "__main__":
    args = make_parser().parse_args()
    input_shape = [int(x) for x in args.shape.split(",")]

    if args.fp16:
        precision_mode = "fp16"
        calibrator = None
    elif args.int8:
        assert (
            args.calibration_data is not None and args.calibration_cache is not None
        ), "calibration data path and cache path must be provided under int8 mode"
        precision_mode = "int8"
        calibrator = MyCalibrator(
            args.calibration_data, 1, (1, *input_shape), args.calibration_cache
        )
    else:
        print("You must enable fp16 or int8!")
        exit(1)

    is_success = export(
        args.onnx,
        args.ouput,
        input_shape,
        precision_mode=precision_mode,
        calibrator=calibrator,
    )

    if is_success:
        print(f"TensorRT engine saved as {args.ouput}")
